pandas/tools/merge.py

```diff
@@ -896,8 +896,18 @@ def get_result(self):
                 data = dict(zip(range(len(self.objs)), self.objs))
                 index, columns = self.new_axes
                 tmpdf = DataFrame(data, index=index)
-                if columns is not None:
-                    tmpdf.columns = columns
+                # checks if the columns names are equal to a list whose elements are a range starting from zero and
+                # ending at len(self.objs)-1. If that's the case, use the series names as column names. This is
+                # necessary to inherit the column names from the 'keys' argument in the 'concat' function call, when
+                # provided
+                if (columns == [range(len(self.objs))]).all():
```

Comment:
I see, that's a much cleaner way to do it. Thanks for sharing that with me @jreback 
You can discard my pull request in this case and go ahead merging your patch. I you like you can keep my fixes for the unit tests.